### DISCUSSION TITLE
lazy/cache method code generation

### DISCUSSION THREAD

**Discussion by AskMeAgain** [mapstruct/mapstruct#2878] [Ideas]
# What is this feature about?

I want to annotate specific methods (even methods which are not directly used by mapstruct) with @Lazy and mapstruct generates automatically a lazy/cache layer for it

At work we have to calculate stuff multiple times via alot of nested calls. We have 1 big input dto, and do alot of mathematical calculations around this dto. We then map this to an output dto. Often the calculations are used multiple times and we could save alot of cpu cycles if we could reuse some of the results.

Writing the lazy layer by ourselves is possible, but clunky. We could save alot of code on our side if mapstruct would supports this feature.

With this feature its possible to use mapstruct as an efficient calculator, while also having a nice and clean code base.

I am willing to implement this feature by myself, but first i want to know if such a feature has a chance to get merged.

# Code example
    
### The mapper

    @Mapper
    public interface LazyMapper {
            
        @Mapping(target = "a", source = "a", qualifiedByName = "lazyMethod")
        @Mapping(target = "b", source = "b", qualifiedByName = "lazyMethod")
        Target mapLazy(Source source);
    
        @Named("lazyMethod")
        default Integer lazyMethod(Integer input) {
            return getLazyStuff(input) + getLazyStuff(input) + getLazyStuff(input) + getLazyStuff(input);
        }
    
        @Lazy
        default Integer getLazyStuff(Integer input){
            //SUPER LONG CALCUATION
            return input;
        }
    }

### The generated code
    
    public class LazyMapperImpl implements LazyMapper {
    
        private final ThreadLocal<Integer> _lazyMethod = new ThreadLocal<>();
    
        @Override
        public Target mapLazy(Source source) {
            if ( source == null ) {
                return null;
            }
    
            Integer a = null;
            Integer b = null;
    
            a = lazyMethod( source.getA() );
            b = lazyMethod( source.getB() );
    
            Target target = new Target( a, b);

            _lazyMethod.remove();
    
            return target;
        }
    
        @Override
        public Integer getLazyStuff(Integer input){
            if (_lazyMethod.get() == null){
                _lazyMethod.set(LazyMapper.super.getLazyStuff(input));
            }
            return _lazyMethod.get();
        }
    }

**Comment by Zegveld** [comment]
I don't get your example.

1. It does not do any caching:
if `a=10`, and `b=20` and the result of calculating `a` would be `13` and `b` would be `26` then your current code would call `new Target( 13, 13 )` because it calculates `a` and then also retrieves this result for `b`.
2. It is listed as lazy, but it is not lazy at all, it immediately executes. 

Concerning caching, I see several options:
* using a structure like the [CycleAvoidingMappingContext](https://github.com/mapstruct/mapstruct-examples/blob/master/mapstruct-mapping-with-cycles/src/main/java/org/mapstruct/example/mapper/CycleAvoidingMappingContext.java).
* Writing a generic caching mechanism in a base class which will be the source of all your calculations.

It is very easy to introduce caching, using your example:
```
    Map<Integer, Integer> resultCache = new HashMap<>();

    @Named("lazyMethod")
    default Integer lazyMethod(Integer input) {
        return resultCache.computeIfAbsent(input, key -> getLazyStuff(input) + getLazyStuff(input) + getLazyStuff(input) + getLazyStuff(input));
    }

```
With changing one line and adding another caching has been introduced for your calculations.

Also terminology as i know it:
* Lazy: used to refer to lazy loading, for example returning a collection that gets populated the moment you start retrieving values from it.
* Cache: Storing a previous result linked to specific input, then return that result if the same input is supplied.

**Comment by AskMeAgain** [comment]
Lazy in this case means, that we calculate it a single time and then get the cached result (we are lazy because we already know the result), but if you want to name it caching, iam fine with it.

Maybe my example should be a bit better, but the idea is just that "getLazyResult(input)" is executed multiple times, but we only need to compute it once and would do the lookup via the threadlocal.

> Writing the lazy layer by ourselves is possible, but clunky. We could save alot of code on our side if mapstruct would supports this feature.

your resultCache solution is alot of boilerplate code, just for caching. A simple annotation would be alot easier to read. Same with the CycleAvoidingMappingContext. 

Iam fine if this feature gets rejected, since its not what mapstruct normally does (mapping), but for me a calculator is also just a mapping of multiple values to a result value, and adding a caching layer around it is a nice feature.

Thanks for your time

**Comment by filiphr** [comment]
I think that I understand what you want to achieve @AskMeAgain.

However, I am not sure that I agree that something like this belongs to MapStruct. This can easily blow out of proportion for how to store the computed value. Should it be in a ThreadLocal, should it be per mapping method, should it be global, what kind of cache should it be and so on.

There are numerous caching libraries in Java that offer a lot of different options.

In my opinion this could be already achieves with the combination of `@BeforeMapping`, `@AfterMapping`. 

What we could do is to support using mapping methods from parameters `@Context`, I think that currently those methods are not used for caching. This would allow you to store the computed value in a normal field, since the context can be instantiated every time you invoke the mapping method.

>Iam fine if this feature gets rejected, since its not what mapstruct normally does (mapping), but for me a calculator is also just a mapping of multiple values to a result value, and adding a caching layer around it is a nice feature.

I do agree with you for the calculator. However, I don't think that we are the right place to add a caching layer for these things. If there are things that we could improve, e.g. not invoke a method twice, or store something in a method variable then we should do. However, this goes a bit outside of the scope of MapStruct.

There are other libraries that would offer something like this. I know that Spring has `@Cacheable` and offers different implementations options for different caches.

**Comment by AskMeAgain** [comment]
> However, I am not sure that I agree that something like this belongs to MapStruct. This can easily blow out of proportion for how to store the computed value. Should it be in a ThreadLocal, should it be per mapping method, should it be global, what kind of cache should it be and so on.

I think this is an implementation detail and in the end doesnt really matter: Threadlocals with a concurrent hashmap would be a sufficient solution.

> There are numerous caching libraries in Java that offer a lot of different options.

Yes, but none can provide an annotation based approach (which i think is best for a super lean code base)

> In my opinion this could be already achieves with the combination of @BeforeMapping, @AfterMapping.

How would you do this with the example above? MapStruct can only use @BeforeMapping/AfterMapping on mapping methods it knows about. getLazyStuff() is used in these methods, but mapstruct cannot know this. 

> What we could do is to support using mapping methods from parameters @Context, I think that currently those methods are not used for caching. This would allow you to store the computed value in a normal field, since the context can be instantiated every time you invoke the mapping method.

This would only allow caching for mapstruct-known-methods. 

Also i think that we can still make the mapping work in a "stateless" way, by cleaning up the cache after a mapping, but again this is just an implementation detail.

> There are other libraries that would offer something like this. I know that Spring has @Cacheable and offers different implementations options for different caches.

The thing is that all the proxy methods dont work in this case, since proxies are only used when you execute methods from outside, and
not from within an interface. So AoP is completely out.

I looked pretty hard for a (good) solution to this problem and the only way i came up with, is using MapStruct + [LazyGen (by me)] (https://github.com/AskMeAgain/LazyGen) but this is also rather clunky, and not without problems, and i would prefer to integrate this into mapstruct directly.

> However, this goes a bit outside of the scope of MapStruct.

Sadly i understand this completely.

Maybe we could think about a plugin architecture where we can provide/extent the existing templates? If you want i can spend more time on this and start a discussion later.
